official/recommendation/ncf_main.py

```diff
@@ -0,0 +1,300 @@
+
+  run_config = tf.estimator.RunConfig(train_distribute=distribution)
+
+  est_model = tf.keras.estimator.model_to_estimator(
+      keras_model=keras_model, model_dir=model_dir, config=run_config)
+
+  return est_model
+
+
+def per_device_batch_size(batch_size, num_gpus):
+  """For multi-gpu, batch-size must be a multiple of the number of GPUs.
+
+  Note that this should eventually be handled by DistributionStrategies
+  directly. Multi-GPU support is currently experimental, however,
+  so doing the work here until that feature is in place.
+
+  Args:
+    batch_size: Global batch size to be divided among devices. This should be
+      equal to num_gpus times the single-GPU batch_size for multi-gpu training.
+    num_gpus: How many GPUs are used with DistributionStrategies.
+
+  Returns:
+    Batch size per device.
+
+  Raises:
+    ValueError: if batch_size is not divisible by number of devices
+  """
+  if num_gpus <= 1:
+    return batch_size
+
+  remainder = batch_size % num_gpus
+  if remainder:
+    err = ("When running with multiple GPUs, batch size "
+           "must be a multiple of the number of available GPUs. Found {} "
+           "GPUs with a batch size of {}; try --batch_size={} instead."
+          ).format(num_gpus, batch_size, batch_size - remainder)
+    raise ValueError(err)
+  return int(batch_size / num_gpus)
+
+
+def main(_):
+  # Data preprocessing
+  # The file name of training and test dataset
+  train_fname = os.path.join(
+      FLAGS.data_dir, FLAGS.dataset + "-" + constants.TRAIN_RATINGS_FILENAME)
+  test_fname = os.path.join(
+      FLAGS.data_dir, FLAGS.dataset + "-" + constants.TEST_RATINGS_FILENAME)
+  neg_fname = os.path.join(
+      FLAGS.data_dir, FLAGS.dataset + "-" + constants.TEST_NEG_FILENAME)
+  t1 = time.time()
+  ncf_dataset = dataset.data_preprocessing(
+      train_fname, test_fname, neg_fname, FLAGS.num_neg)
+  tf.logging.info("Data preprocessing: {:.1f} s".format(time.time() - t1))
+
+  # Create NeuMF model and convert it to Estimator
+  tf.logging.info("Creating Estimator from Keras model...")
+  keras_model = neumf_model.NeuMF(
+      ncf_dataset.num_users, ncf_dataset.num_items, FLAGS.num_factors,
+      ast.literal_eval(FLAGS.layers), FLAGS.batch_size, FLAGS.reg_mf)
+  num_gpus = get_num_gpus(FLAGS.num_gpus)
+  est_model = convert_keras_to_estimator(keras_model, num_gpus, FLAGS.model_dir)
+
+  # Training and evaluation cycle
+  def train_input_fn():
+    return dataset.input_fn(
+        True, per_device_batch_size(FLAGS.batch_size, num_gpus),
+        FLAGS.epochs_between_evals, ncf_dataset, FLAGS.num_parallel_calls)
+
+  total_training_cycle = (FLAGS.train_epochs //
+                          FLAGS.epochs_between_evals)
+  for cycle_index in range(total_training_cycle):
+    tf.logging.info("Starting a training cycle: {}/{}".format(
+        cycle_index, total_training_cycle - 1))
+
+    # Train the model
+    train_cycle_begin = time.time()
```

Comment:
Can be done in the followup, but probably not necessary to collect/print out this very rough timing info.